enterprise/internal/notebooks/store.go

```
@@ -0,0 +1,202 @@
+	"encoding/json"
+
+	"github.com/cockroachdb/errors"
+
+	"github.com/keegancsmith/sqlf"
+
+	"github.com/sourcegraph/sourcegraph/internal/actor"
+	"github.com/sourcegraph/sourcegraph/internal/database/basestore"
+	"github.com/sourcegraph/sourcegraph/internal/database/dbutil"
+)
+
+var ErrNotebookNotFound = errors.New("notebook not found")
+
+type NotebooksOrderByOption uint8
+
+const (
+	NotebooksOrderByID NotebooksOrderByOption = iota
+	NotebooksOrderByUpdatedAt
+	NotebooksOrderByCreatedAt
+)
+
+type ListNotebooksPageOptions struct {
+	First int32
+	After int32
+}
+
+type ListNotebooksOptions struct {
+	Query             string
+	OrderBy           NotebooksOrderByOption
+	OrderByDescending bool
+}
+
+func Notebooks(db dbutil.DB) NotebooksStore {
+	store := basestore.NewWithDB(db, sql.TxOptions{})
+	return &notebooksStore{store}
+}
+
+type NotebooksStore interface {
+	basestore.ShareableStore
+	GetNotebook(context.Context, int64) (*Notebook, error)
+	CreateNotebook(context.Context, *Notebook) (*Notebook, error)
+	// TODO
+	// UpdateNotebook(context.Context, *Notebook) (*Notebook, error)
+	// DeleteNotebook(context.Context, int64) (*Notebook, error)
+	// ListNotebooks(context.Context, ListNotebooksPageOptions, ListNotebooksOptions) ([]*Notebook, error)
+	// CountNotebooks(context.Context, ListNotebooksOptions) (int, error)
+}
+
+type notebooksStore struct {
+	*basestore.Store
+}
+
+func (s *notebooksStore) Transact(ctx context.Context) (*notebooksStore, error) {
+	txBase, err := s.Store.Transact(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &notebooksStore{Store: txBase}, nil
+}
+
+const notebooksPermissionsConditionFmtStr = `(
+    -- Bypass permission check
+    %s
+    -- Happy path of public notebooks
+    OR n.public
+    -- Private notebooks are available only to its creator
+    OR (n.creator_user_id IS NOT NULL AND n.creator_user_id = %d)
+)`
+
+func notebooksPermissionsCondition(ctx context.Context, db dbutil.DB) (*sqlf.Query, error) {
```

nit: does this function need to return an error?

---

Nope, I'll simplify.